execution_engine_testing/tests/src/test/chainspec_registry.rs

```diff
@@ -0,0 +1,243 @@
+    builder.run_genesis(&run_genesis_request);
+}
+
+#[ignore]
+#[test]
+#[should_panic]
+fn should_fail_to_commit_genesis_when_missing_genesis_accounts_hash() {
+    let mut rng = rand::thread_rng();
+    let chainspec_bytes_hash = Digest::hash(rng.gen::<[u8; 32]>());
+
+    let mut incomplete_chainspec_registry = ChainspecRegistry::new();
+    incomplete_chainspec_registry.insert(CHAINSPEC_RAW.to_string(), chainspec_bytes_hash);
+
+    let run_genesis_request = RunGenesisRequest::new(
+        *DEFAULT_GENESIS_CONFIG_HASH,
+        *DEFAULT_PROTOCOL_VERSION,
+        DEFAULT_EXEC_CONFIG.clone(),
+        incomplete_chainspec_registry.clone(),
+    );
+
+    let mut builder = InMemoryWasmTestBuilder::default();
+    builder.run_genesis(&run_genesis_request);
+}
+
+#[ignore]
+#[test]
+fn should_write_chainspec_registry_during_an_upgrade() {
+    let mut rng = rand::thread_rng();
+
+    let mut builder = InMemoryWasmTestBuilder::default();
+    builder.run_genesis(&DEFAULT_RUN_GENESIS_REQUEST);
+
+    let mut upgraded_chainspec_registry = ChainspecRegistry::new();
+    let chainspec_bytes_hash = Digest::hash(rng.gen::<[u8; 32]>());
+    upgraded_chainspec_registry.insert(CHAINSPEC_RAW.to_string(), chainspec_bytes_hash);
+    let global_state_toml_hash = Digest::hash(rng.gen::<[u8; 32]>());
+    upgraded_chainspec_registry.insert(GLOBAL_STATE_RAW.to_string(), global_state_toml_hash);
+
+    let mut upgrade_request = {
+        UpgradeRequestBuilder::new()
+            .with_current_protocol_version(*OLD_PROTOCOL_VERSION)
+            .with_new_protocol_version(*NEW_PROTOCOL_VERSION)
+            .with_activation_point(DEFAULT_ACTIVATION_POINT)
+            .with_chainspec_registry(upgraded_chainspec_registry)
+            .build()
+    };
+
+    let engine_config = EngineConfig::default();
+
+    builder
+        .upgrade_with_upgrade_request(engine_config, &mut upgrade_request)
+        .expect_upgrade_success();
+
+    let queried_registry = builder
+        .query(None, Key::ChainspecRegistry, &[])
+        .expect("must have entry under Key::ChainspecRegistry")
+        .as_cl_value()
+        .expect("must have underlying cl_value")
+        .to_owned()
+        .into_t::<ChainspecRegistry>()
+        .expect("must convert to chainspec registry");
+
+    // There should be no entry for the genesis accounts once the upgrade has completed.
+    assert!(queried_registry.get(GENESIS_ACCOUNTS_RAW).is_none());
+
+    let queried_chainspec_hash = queried_registry
+        .get(CHAINSPEC_RAW)
+        .expect("must have entry for chainspec_hash");
+
+    assert_eq!(*queried_chainspec_hash, chainspec_bytes_hash);
+
+    let queried_global_state_toml_hash = queried_registry
+        .get(GLOBAL_STATE_RAW)
+        .expect("must have entry for genesis accounts");
+
+    assert_eq!(*queried_global_state_toml_hash, global_state_toml_hash);
+}
+
+#[ignore]
+#[test]
+fn should_upgrade_and_write_registry_from_release_1_4_4() {
```

Comment:
The code in this function is like 98% copy&pasted from `should_write_chainspec_registry_during_an_upgrade()`. Please consider some dedup.

Reply:
Yeah, thanks for pointing this out, I've collated the code into one function now